core/src/main/java/nu/yona/server/subscriptions/service/UserService.java

```diff
@@ -479,45 +474,52 @@ User addUserCreatedOnBuddyRequest(UserDto buddyUserResource)
 	@Transactional
 	public UserDto updateUser(UUID id, UserDto user)
 	{
-		User originalUserEntity = getUserEntityById(id);
-		UserDto originalUser = createUserDtoWithPrivateData(originalUserEntity);
-		assertValidUpdateRequest(user, originalUser, originalUserEntity);
+		User updatedUserEntity = updateUser(id, userEntity -> {
+			UserDto originalUser = createUserDtoWithPrivateData(userEntity);
+			assertValidUpdateRequest(user, originalUser, userEntity);
 
-		User updatedUserEntity = user.updateUser(originalUserEntity);
-		Optional<ConfirmationCode> confirmationCode = createConfirmationCodeIfNumberUpdated(user, originalUser,
-				updatedUserEntity);
-		User savedUserEntity = userRepository.save(updatedUserEntity);
-		if (confirmationCode.isPresent())
-		{
-			sendConfirmationCodeTextMessage(updatedUserEntity.getMobileNumber(), confirmationCode.get(),
-					SmsTemplate.CHANGED_USER_NUMBER_CONFIRMATION);
-		}
-		UserDto userDto = createUserDtoWithPrivateData(savedUserEntity);
-		logger.info("Updated user with mobile number '{}' and ID '{}'", userDto.getMobileNumber(), userDto.getId());
-		buddyService.broadcastUserInfoChangeToBuddies(savedUserEntity, originalUser);
-		return userDto;
+			user.updateUser(userEntity);
+			Optional<ConfirmationCode> confirmationCode = createConfirmationCodeIfNumberUpdated(user, originalUser, userEntity);
+			if (confirmationCode.isPresent())
+			{
+				sendConfirmationCodeTextMessage(userEntity.getMobileNumber(), confirmationCode.get(),
+						SmsTemplate.CHANGED_USER_NUMBER_CONFIRMATION);
+			}
+			UserDto userDto = createUserDtoWithPrivateData(userEntity);
+			logger.info("Updated user with mobile number '{}' and ID '{}'", userDto.getMobileNumber(), userDto.getId());
+			buddyService.broadcastUserInfoChangeToBuddies(userEntity, originalUser);
+		});
+		return createUserDtoWithPrivateData(updatedUserEntity);
+	}
+
+	@Transactional
+	public User updateUser(UUID id, Consumer<User> updateAction)
+	{
+		User user = getUserEntityByIdWithUpdateLock(id);
```

Comment:
Please add a comment here that we do this locking because read and updates on the `User` are done in the same request for special purposes (upgrade mechanism).

Reply:
Done.